jena-arq/src/main/java/org/apache/jena/riot/system/ErrorHandlerFactory.java

```diff
@@ -221,6 +224,64 @@ public void fatal(String message, long line, long col) {
         }
     }
     
+    /** An error handler that logs message for errors and warnings and throw exceptions on either */ 
+    public static class ErrorHandlerTracking extends ErrorLogger implements ErrorHandler {
+        private final boolean failOnError, failOnWarning;
+        private long errorCount, warningCount;
+        
+        public ErrorHandlerTracking(Logger log, boolean failOnError, boolean failOnWarning) {
```

Comment:
Suggestion: `ErrorHandlerTracking` wraps another `ErrorHandler` as argument and forwards all warning/error/fatal calls to that error handler after incrementing its counters.  This makes the choice of how to respond to warning or error dependent on the argument `ErrorHandler`.